cocotb/bfmgen.py

```diff
@@ -0,0 +1,398 @@
+        bfm_import_calls += "                  " + imp.T.__name__  + "(\n"
+        for pi in range(len(imp.signature)):
+            p = imp.signature[pi]
+            if p.ptype.s:
+                bfm_import_calls += "                      $cocotb_bfm_get_param_i32(bfm_id)"
+            else:
+                bfm_import_calls += "                      $cocotb_bfm_get_param_ui32(bfm_id)"
+            
+            if pi+1 < len(imp.signature):
+                bfm_import_calls += ","
+            bfm_import_calls += "\n"
+        bfm_import_calls += "                      );\n"
+        bfm_import_calls += "              end\n"
+        
+    bfm_export_tasks = ""
+    for i in range(len(info.export_info)):
+        exp = info.export_info[i]
+        bfm_export_tasks += "    task " + exp.T.__name__ 
+        
+        if len(exp.signature) > 0:
+            bfm_export_tasks += "("
+            for j in range(len(exp.signature)):
+                p = exp.signature[j]
+                bfm_export_tasks += p.ptype.vl_type() + " " + p.pname
+                if j+1 < len(exp.signature):
+                    bfm_export_tasks += ", "
+            bfm_export_tasks += ");\n"
+        else:
+            bfm_export_tasks += ";\n"
+        bfm_export_tasks += "    begin\n"
+        bfm_export_tasks += "        $cocotb_bfm_begin_msg(bfm_id, " + str(i) + ");\n"
+        for p in exp.signature:
+            if p.ptype.s:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_si(bfm_id, " + p.pname + ");\n"
+            else:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_ui(bfm_id, " + p.pname + ");\n"
+            
+        bfm_export_tasks += "        $cocotb_bfm_end_msg(bfm_id);\n"
+        bfm_export_tasks += "    end\n"
+        bfm_export_tasks += "    endtask\n"
+        
+    
+    impl_param_m = {
+        "bfm_classname" : info.T.__module__ + "." + info.T.__qualname__,
+        "bfm_import_calls" : bfm_import_calls,
+        "bfm_export_tasks" : bfm_export_tasks
+        }
+    
+    cocotb_bfm_api_impl = '''
+    reg signed[31:0]      bfm_id;
+    event                 bfm_ev;
+    reg signed[31:0]      bfm_msg_id;
+    
+${bfm_export_tasks}
+    
+    initial begin
+      bfm_id = $cocotb_bfm_register("${bfm_classname}", bfm_ev);
+      
+      while (1) begin
+          bfm_msg_id = $cocotb_bfm_claim_msg(bfm_id);
+          
+          case (bfm_msg_id)
+${bfm_import_calls}
+              -1: begin
+                  @(bfm_ev);
+              end
+          endcase
+          
+      end
+    end
+    '''
+   
+    param_m = {
+        "cocotb_bfm_api_impl" : Template(cocotb_bfm_api_impl).safe_substitute(impl_param_m)
+        }
+    
+    
+    return t.safe_substitute(param_m)
+
+def bfm_generate_vl(args):
+    inst = BfmMgr.inst()
+    
+    out = open(args.o, "w")
+    out.write("//***************************************************************************\n")
+    out.write("//* BFMs file for CocoTB. \n")
```

Comment:
Nit: I have tried to normalize the spelling of "cocotb" to be all lowercase in the last few months.
(I think we are still inconsistent when it's at the beginning of a sentence though.)

Reply:
Perfect, update on the way